File: web/src/displayList.ts
export type Stop = { at: number; color: Float32Array }
export type Paint =
  | { type: 'solid'; color: Float32Array }
  | { type: 'linear' | 'radial'; transform: Float32Array; stops: Stop[] }
export type Shadow = { offset: Float32Array; blur: number; color: Float32Array }

export type Op =
  | { op: 'page'; width: number; height: number; bleed: number }
  | { op: 'beginItem'; item: number; hash: number }
  | { op: 'endItem' }
  | { op: 'fillPath'; paint: Paint; path: Float32Array }
  | {
      op: 'glyphRun'
      font: number
      size: number
      paint: Paint
      glyphs: Uint16Array
      positions: Float32Array
    }
  | { op: 'image'; image: number; rect: Float32Array }
  | { op: 'pushClip'; path: Float32Array; invert: boolean }
  | { op: 'popClip' }
  | { op: 'strokePath'; paint: Paint; width: number; cap: number; join: number; path: Float32Array }
  | { op: 'pushLayer'; hash: number; opacity: number; blend: number; blur: number; shadows: Shadow[] }
  | { op: 'popLayer' | 'beginMask' | 'endMask' | 'popMask' }

const SIMPLE = { 2: 'endItem', 7: 'popClip', 10: 'popLayer', 11: 'beginMask', 12: 'endMask', 13: 'popMask' } as const

export function decode(words: Uint32Array): Op[] {
  const { buffer, byteOffset } = words
  const f32 = (at: number, n: number) => new Float32Array(buffer, byteOffset + at * 4, n)
  const ops: Op[] = []
  let i = 0
  const paint = (): Paint => {
    const type = words[i++]
    if (type === 0) {
      i += 4
      return { type: 'solid', color: f32(i - 4, 4) }
    }
    const transform = f32(i, 6)
    const n = words[i + 6]
    i += 7
    const stops = Array.from({ length: n }, (_, k) => ({ at: f32(i + 5 * k, 1)[0], color: f32(i + 5 * k + 1, 4) }))
    i += 5 * n
    return { type: type === 1 ? 'linear' : 'radial', transform, stops }
  }
  const path = () => {
    const n = words[i]
    i += 1 + n
    return f32(i - n, n)
  }
  while (i < words.length) {
    const code = words[i++]
    switch (code) {
      case 0: {
        const [width, height, bleed] = f32(i, 3)
        ops.push({ op: 'page', width, height, bleed })
        i += 3
        break
      }
      case 1:
        ops.push({ op: 'beginItem', item: words[i], hash: words[i + 1] })
        i += 2
        break
      case 3: {
        const p = paint()
        ops.push({ op: 'fillPath', paint: p, path: path() })
        break
      }
      case 4: {
        const font = words[i]
        const size = f32(i + 1, 1)[0]
        i += 2
        const p = paint()
        const n = words[i]
        const glyphs = new Uint16Array(buffer, byteOffset + (i + 1) * 4, n)
        i += 1 + Math.ceil(n / 2)
        ops.push({ op: 'glyphRun', font, size, paint: p, glyphs, positions: f32(i, 2 * n) })
        i += 2 * n
        break
      }
      case 5:
        ops.push({ op: 'image', image: words[i], rect: f32(i + 1, 4) })
        i += 5
        break
      case 6: {
        const invert = words[i++] === 1
        ops.push({ op: 'pushClip', invert, path: path() })
        break
      }
      case 8: {
        const p = paint()
        const width = f32(i, 1)[0]
        const [cap, join] = [words[i + 1], words[i + 2]]
        i += 3
        ops.push({ op: 'strokePath', paint: p, width, cap, join, path: path() })
        break
      }
      case 9: {
        const hash = words[i]
        const [opacity] = f32(i + 1, 1)
        const blend = words[i + 2]
        const [blur] = f32(i + 3, 1)
        const n = words[i + 4]
        i += 5
        const shadows = Array.from({ length: n }, (_, k) => ({
          offset: f32(i + 7 * k, 2),
          blur: f32(i + 7 * k + 2, 1)[0],
          color: f32(i + 7 * k + 3, 4),
        }))
        i += 7 * n
        ops.push({ op: 'pushLayer', hash, opacity, blend, blur, shadows })
        break
      }
      default:
        if (!(code in SIMPLE)) throw new Error(`display list: bad op ${code} at word ${i - 1}`)
        ops.push({ op: SIMPLE[code as keyof typeof SIMPLE] })
    }
  }
  return ops
}

const OPEN = new Set(['pushClip', 'pushLayer', 'beginMask', 'beginItem'])
const CLOSE = new Set(['popClip', 'popLayer', 'popMask', 'endItem', 'endMask'])

/** Index of the op that closes the one opened at `at`; `endMask` closes `beginMask` and opens up to `popMask`. */
export function close(ops: Op[], at: number) {
  let depth = 1
  for (let i = at + 1; i < ops.length; i++) {
    const op = ops[i].op
    if (CLOSE.has(op) && --depth === 0) return i
    if (OPEN.has(op) || op === 'endMask') depth++
  }
  return ops.length
}

File: engine/src/display_list.rs
use serde::Serialize;

pub const MOVE: f32 = 0.0;
pub const LINE: f32 = 1.0;
pub const CUBIC: f32 = 4.0;
pub const CLOSE: f32 = 5.0;

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "op", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum Op {
    Page {
        width: f32,
        height: f32,
        bleed: f32,
    },
    BeginItem {
        item: u32,
    },
    EndItem,
    FillPath {
        paint: Paint,
        path: Vec<f32>,
    },
    GlyphRun {
        font: u32,
        size: f32,
        paint: Paint,
        glyphs: Vec<u16>,
        positions: Vec<f32>,
    },
    Image {
        image: u32,
        rect: [f32; 4],
    },
    PushClip {
        path: Vec<f32>,
        invert: bool,
    },
    PopClip,
    StrokePath {
        paint: Paint,
        width: f32,
        cap: u32,
        join: u32,
        path: Vec<f32>,
    },
    /// Blur and shadow blurs are Gaussian sigmas in pt.
    PushLayer {
        opacity: f32,
        blend: u32,
        blur: f32,
        shadows: Vec<Shadow>,
    },
    PopLayer,
    /// The ops up to `EndMask` are an alpha mask for the ops up to `PopMask`.
    BeginMask,
    EndMask,
    PopMask,
}

/// Gradients map their unit space into page space with `transform` [a b c d e f]:
/// linear runs from (0, 0) to (1, 0), radial is the unit circle around (0, 0).
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Paint {
    Solid {
        color: [f32; 4],
    },
    Linear {
        transform: [f32; 6],
        stops: Vec<Stop>,
    },
    Radial {
        transform: [f32; 6],
        stops: Vec<Stop>,
    },
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Stop {
    pub at: f32,
    pub color: [f32; 4],
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Shadow {
    pub offset: [f32; 2],
    pub blur: f32,
    pub color: [f32; 4],
}

pub fn rect(x: f32, y: f32, w: f32, h: f32) -> Vec<f32> {
    vec![
        MOVE,
        x,
        y,
        LINE,
        x + w,
        y,
        LINE,
        x + w,
        y + h,
        LINE,
        x,
        y + h,
        CLOSE,
    ]
}

fn floats(out: &mut Vec<u32>, v: &[f32]) {
    out.extend(v.iter().map(|f| f.to_bits()));
}

fn paint(out: &mut Vec<u32>, p: &Paint) {
    match p {
        Paint::Solid { color } => {
            out.push(0);
            floats(out, color);
        }
        Paint::Linear { transform, stops } | Paint::Radial { transform, stops } => {
            out.push(if matches!(p, Paint::Linear { .. }) {
                1
            } else {
                2
            });
            floats(out, transform);
            out.push(stops.len() as u32);
            for s in stops {
                floats(out, &[s.at]);
                floats(out, &s.color);
            }
        }
    }
}

/// Index of the op that closes the one opened at `at`.
/// `EndMask` closes `BeginMask` and opens up to `PopMask`.
pub fn close(ops: &[Op], at: usize) -> usize {
    let mut depth = 1;
    for (i, op) in ops.iter().enumerate().skip(at + 1) {
        match op {
            Op::PopClip | Op::PopLayer | Op::PopMask | Op::EndItem | Op::EndMask => {
                depth -= 1;
                if depth == 0 {
                    return i;
                }
                if *op == Op::EndMask {
                    depth += 1;
                }
            }
            Op::PushClip { .. } | Op::PushLayer { .. } | Op::BeginMask | Op::BeginItem { .. } => {
                depth += 1
            }
            _ => {}
        }
    }
    ops.len()
}

/// Writes the hash of the words since the innermost open item or layer into the word before them.
fn seal(out: &mut [u32], open: &mut Vec<usize>) {
    let start = open.pop().expect("close without open");
    out[start - 1] = out[start..]
        .iter()
        .fold(0x811c9dc5, |h, w| (h ^ w).wrapping_mul(0x01000193));
}

pub fn encode(ops: &[Op]) -> Vec<u32> {
    let mut out = Vec::new();
    let mut open = Vec::new();
    for op in ops {
        match op {
            Op::Page {
                width,
                height,
                bleed,
            } => {
                out.push(0);
                floats(&mut out, &[*width, *height, *bleed]);
            }
            Op::BeginItem { item } => {
                out.extend([1, *item, 0]);
                open.push(out.len());
            }
            Op::EndItem => {
                seal(&mut out, &mut open);
                out.push(2);
            }
            Op::FillPath { paint: p, path } => {
                out.push(3);
                paint(&mut out, p);
                out.push(path.len() as u32);
                floats(&mut out, path);
            }
            Op::GlyphRun {
                font,
                size,
                paint: p,
                glyphs,
                positions,
            } => {
                out.extend([4, *font, size.to_bits()]);
                paint(&mut out, p);
                out.push(glyphs.len() as u32);
                out.extend(
                    glyphs
                        .chunks(2)
                        .map(|c| c[0] as u32 | (*c.get(1).unwrap_or(&0) as u32) << 16),
                );
                floats(&mut out, positions);
            }
            Op::Image { image, rect } => {
                out.extend([5, *image]);
                floats(&mut out, rect);
            }
            Op::PushClip { path, invert } => {
                out.extend([6, *invert as u32, path.len() as u32]);
                floats(&mut out, path);
            }
            Op::PopClip => out.push(7),
            Op::StrokePath {
                paint: p,
                width,
                cap,
                join,
                path,
            } => {
                out.push(8);
                paint(&mut out, p);
                out.extend([width.to_bits(), *cap, *join, path.len() as u32]);
                floats(&mut out, path);
            }
            Op::PushLayer {
                opacity,
                blend,
                blur,
                shadows,
            } => {
                out.extend([9, 0]);
                open.push(out.len());
                out.extend([opacity.to_bits(), *blend, blur.to_bits()]);
                out.push(shadows.len() as u32);
                for s in shadows {
                    floats(&mut out, &s.offset);
                    floats(&mut out, &[s.blur]);
                    floats(&mut out, &s.color);
                }
            }
            Op::PopLayer => {
                seal(&mut out, &mut open);
                out.push(10);
            }
            Op::BeginMask => out.push(11),
            Op::EndMask => out.push(12),
            Op::PopMask => out.push(13),
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn item_hash_follows_item_content() {
        let hash = |color| {
            encode(&[
                Op::BeginItem { item: 0 },
                Op::FillPath {
                    paint: Paint::Solid { color },
                    path: rect(0.0, 0.0, 1.0, 1.0),
                },
                Op::EndItem,
            ])[2]
        };
        assert_eq!(hash([1.0; 4]), hash([1.0; 4]));
        assert_ne!(hash([1.0; 4]), hash([0.5; 4]));
    }

    #[test]
    fn layer_hash_follows_everything_inside_the_layer() {
        let hash = |opacity| {
            encode(&[
                Op::PushLayer {
                    opacity: 1.0,
                    blend: 0,
                    blur: 0.0,
                    shadows: vec![],
                },
                Op::PushLayer {
                    opacity,
                    blend: 0,
                    blur: 0.0,
                    shadows: vec![],
                },
                Op::PopLayer,
                Op::PopLayer,
            ])[1]
        };
        assert_eq!(hash(0.5), hash(0.5));
        assert_ne!(hash(0.5), hash(0.25));
    }

    #[test]
    fn close_finds_the_matching_end() {
        let ops = [
            Op::BeginMask,
            Op::BeginItem { item: 0 },
            Op::EndItem,
            Op::EndMask,
            Op::BeginItem { item: 0 },
            Op::EndItem,
            Op::PopMask,
        ];
        assert_eq!([0, 3, 1].map(|at| close(&ops, at)), [3, 6, 2]);
    }

    #[test]
    fn writes_fixture_for_the_ts_decoder() {
        let ops = vec![
            Op::Page {
                width: 420.5,
                height: 595.25,
                bleed: 8.5,
            },
            Op::PushClip {
                path: rect(1.0, 2.0, 3.0, 4.0),
                invert: true,
            },
            Op::PushLayer {
                opacity: 0.5,
                blend: 1,
                blur: 2.0,
                shadows: vec![Shadow {
                    offset: [1.0, 2.0],
                    blur: 3.0,
                    color: [0.0, 0.0, 0.0, 0.25],
                }],
            },
            Op::BeginMask,
            Op::FillPath {
                paint: Paint::Solid {
                    color: [0.0, 0.0, 0.0, 1.0],
                },
                path: rect(0.0, 0.0, 1.0, 1.0),
            },
            Op::EndMask,
            Op::BeginItem { item: 7 },
            Op::FillPath {
                paint: Paint::Linear {
                    transform: [1.0, 0.0, 0.0, 1.0, 5.0, 6.0],
                    stops: vec![
                        Stop {
                            at: 0.0,
                            color: [1.0, 0.5, 0.25, 1.0],
                        },
                        Stop {
                            at: 1.0,
                            color: [0.0, 0.5, 1.0, 0.5],
                        },
                    ],
                },
                path: rect(10.0, 20.0, 30.5, 40.0),
            },
            Op::StrokePath {
                paint: Paint::Radial {
                    transform: [2.0, 0.0, 0.0, 3.0, 1.0, 1.0],
                    stops: vec![Stop {
                        at: 0.5,
                        color: [1.0, 1.0, 1.0, 1.0],
                    }],
                },
                width: 1.5,
                cap: 1,
                join: 2,
                path: rect(1.0, 1.0, 2.0, 2.0),
            },
            Op::GlyphRun {
                font: 0,
                size: 12.0,
                paint: Paint::Solid {
                    color: [0.0, 0.0, 0.0, 1.0],
                },
                glyphs: vec![3, 65535, 42],
                positions: vec![0.0, 0.0, 6.5, 0.0, 13.0, 0.0],
            },
            Op::Image {
                image: 2,
                rect: [1.0, 2.0, 3.0, 4.0],
            },
            Op::EndItem,
            Op::PopMask,
            Op::PopLayer,
            Op::PopClip,
        ];
        let words = encode(&ops);
        let bytes: Vec<u8> = words.iter().flat_map(|w| w.to_le_bytes()).collect();
        let dir = concat!(env!("CARGO_MANIFEST_DIR"), "/../web/src/testdata");
        std::fs::create_dir_all(dir).unwrap();
        std::fs::write(format!("{dir}/display-list.bin"), bytes).unwrap();
        std::fs::write(
            format!("{dir}/display-list.json"),
            serde_json::to_string_pretty(&ops).unwrap(),
        )
        .unwrap();
    }
}
